### DISCUSSION TITLE
Add dynamic client-side routing to exported sites?

### DISCUSSION THREAD

**Discussion by miloofcroton** [vercel/next.js#12121] [Help]
Say you go to /items/4. There *can* be a 4.html file under /items, but it has to know that exists at the time that you build the site. If it’s a blog, fine. If it’s an app with a non-trivial data model and API, you don’t want to have to keep doing lengthy rebuilds for new items. You just have an [id].js file. This works on the node server because it will dynamically generate 4.html on request. But if you statically export the project, there’s currently nothing in the client that will generate the 4.html. It just errors.

The way vanilla React does it is that you have a router object in JSX, and you pass in components as children that will mount if the route path matches what you give it. This is how client-side routing works. I heard there might be some long-term plan to support react-router, but I think this is the wrong approach.

If Next could take their filename routing with dynamic routing and add in some client-side runtime that is enabled if you export the site, that would take this project to the next level. Would that be hard to do?

Basically, if you export a dynamic route that has a new lifecycle method (`getClientProps`?), it can go the parent, add a switch that routes to said parent html file (if it exists) or the dynamic route if the pattern matches. Basically an isomorphic next-router, I would think.

**Comment by timneutkens** [comment]
Next already supports this pattern through the automatic static optimization. When you don't export `getStaticProps` or `getStaticPaths` then the dynamic route is served as a static file and the parameters are exposed client-side.

This is supported with `next start`. If you use `next export` (I guess that's what you were referring to) you have to set up your proxy/file server to know about dynamic routes. Eg if you have `pages/[username].js` the `out` directory after `next export` will have `out/[username].html`. Unfortunately we can't automatically configure static file serving as that differs on each platform.

If you host on https://vercel.com it's automatically configured for you as it integrates with Next.js.

**Reply by miloofcroton** [reply]
> Eg if you have pages/[username].js the out directory after next export will have out/[username].html. Unfortunately we can't automatically configure static file serving as that differs on each platform

All I'm suggesting is that instead of exporting `/[username].html`, you add a client-side route to `/`, which will check for a pathname match to `[username]` at which point it loads the [username].js file and mounts it in the app.

**Reply by timneutkens** [reply]
That means all routes would need to be routed to index.html otherwise it won't work, which is not ideal in almost all real-world use cases (eg shipping too much code to all your users).

**Reply by miloofcroton** [reply]
Right, that would just equal what CRA can do. You don't think a hybrid approach would be possible? If a page has a `getClientProps` or some other new lifecycle name, then it gets routed to that index.html, otherwise everything stays as is. 

This wouldn't have to be the root index.html as is the case with SPAs either. Say you have `pages/users/[username]`. That route would go to users/index.html rather than the index.html for the entire site.

**Reply by timneutkens** [reply]
> This wouldn't have to be the root index.html as is the case with SPAs either. Say you have pages/users/[username]. That route would go to users/index.html rather than the index.html for the entire site.

Then you still need to configure your proxy/file-server to be aware of it. You get the best outcome if route the files as said through converting the dynamic paths, not hoisted to a top level route as that results in too much code being loaded or too little code being preloaded. Either of those are bad for the performance of your application.

**Reply by miloofcroton** [reply]
> > This wouldn't have to be the root index.html as is the case with SPAs either. Say you have pages/users/[username]. That route would go to users/index.html rather than the index.html for the entire site.
> 
> Then you still need to configure your proxy/file-server to be aware of it. You get the best outcome if route the files as said through converting the dynamic paths, not hoisted to a top level route as that results in too much code being loaded or too little code being preloaded. Either of those are bad for the performance of your application.

Why would this have to happen at the proxy server?

I'm talking about client routes. The server only has to know about the index file. It's exactly how React Router works.

**Reply by jayenashar** [reply]
> Next already supports this pattern through the automatic static optimization. When you don't export `getStaticProps` or `getStaticPaths` then the dynamic route is served as a static file and the parameters are exposed client-side.
> 
> This is supported with `next start`. If you use `next export` (I guess that's what you were referring to) you have to set up your proxy/file server to know about dynamic routes. Eg if you have `pages/[username].js` the `out` directory after `next export` will have `out/[username].html`.

@timneutkens is this documented somewhere?  I'd like to know if/when this behaviour changes.

Are there any examples where someone has a script to create apache .htaccess files or nginx location blocks?  (or any other popular web server)

**Reply by hunkydoryrepair** [reply]
sorry but I am using vercel.com to host and using `output: 'export'` fails on any dynamic routes.
Now, I can just remove the output specification, but if vercel can't handle the static exports, what would?

at least generating a `.htaccess` automatically would be useful. If using IIS people can adapt the .htaccess to what is needed.
